modules/graph/src/graph/sbom/spdx/mod.rs

```diff
@@ -0,0 +1,200 @@
+                            rel,
+                            right.try_into()?,
+                            &tx,
+                        )
+                        .await?
+                    }
+                }
+            }
+        }
+
+        log::info!("Package cache: {lookup_cache:?}");
+
+        Ok(())
+    }
+}
+
+pub struct SpdxRelationship<'spdx>(&'spdx str, &'spdx RelationshipType, &'spdx str);
+
+impl<'spdx> TryFrom<SpdxRelationship<'spdx>> for (&'spdx str, Relationship, &'spdx str) {
+    type Error = ();
+
+    fn try_from(
+        SpdxRelationship(left, rel, right): SpdxRelationship<'spdx>,
+    ) -> Result<Self, Self::Error> {
+        match rel {
+            RelationshipType::Contains => Ok((right, Relationship::ContainedBy, left)),
+            RelationshipType::ContainedBy => Ok((left, Relationship::ContainedBy, right)),
+            RelationshipType::DependsOn => Ok((right, Relationship::DependencyOf, left)),
+            RelationshipType::DependencyOf => Ok((left, Relationship::DependencyOf, right)),
+            RelationshipType::DevDependencyOf => Ok((left, Relationship::DevDependencyOf, right)),
+            RelationshipType::OptionalDependencyOf => {
+                Ok((left, Relationship::OptionalDependencyOf, right))
+            }
+            RelationshipType::ProvidedDependencyOf => {
+                Ok((left, Relationship::ProvidedDependencyOf, right))
+            }
+            RelationshipType::TestDependencyOf => Ok((left, Relationship::TestDependencyOf, right)),
+            RelationshipType::RuntimeDependencyOf => {
+                Ok((left, Relationship::RuntimeDependencyOf, right))
+            }
+            RelationshipType::ExampleOf => Ok((left, Relationship::ExampleOf, right)),
+            RelationshipType::Generates => Ok((right, Relationship::GeneratedFrom, left)),
+            RelationshipType::GeneratedFrom => Ok((left, Relationship::GeneratedFrom, right)),
+            RelationshipType::AncestorOf => Ok((left, Relationship::AncestorOf, right)),
+            RelationshipType::DescendantOf => Ok((right, Relationship::AncestorOf, left)),
+            RelationshipType::VariantOf => Ok((left, Relationship::VariantOf, right)),
+            RelationshipType::BuildToolOf => Ok((left, Relationship::BuildToolOf, right)),
+            RelationshipType::DevToolOf => Ok((left, Relationship::DevToolOf, right)),
+            _ => Err(()),
+        }
+    }
+}
+
+/// Check the document for invalid SPDX license expressions and replace them with `NOASSERTION`.
+pub fn fix_license(mut json: Value) -> (Value, bool) {
+    let mut changed = false;
+    if let Some(packages) = json["packages"].as_array_mut() {
+        for package in packages {
+            if let Some(declared) = package["licenseDeclared"].as_str() {
+                if let Err(err) = spdx_expression::SpdxExpression::parse(declared) {
+                    log::warn!("Replacing faulty SPDX license expression with NOASSERTION: {err}");
```

Comment:
Future: pass in some identifying context so that if we log a warning, folks can track down which SBOM easily.